### PR TITLE
feat: Added precedence of CLI parameters over envs

### LINKED ISSUE #3188: Bug: Arguments precedence inconsistency in CLI `run` command
### Description

For most of the options env file has higher priority but there are some for which CLI options are "superior".

```python
    reload_dirs = env.reload_dirs or reload_dir
    reload_include = env.reload_include or reload_include
    reload_exclude = env.reload_exclude or reload_exclude
    ...
    ssl_certfile = ssl_certfile or env.certfile_path
    ssl_keyfile = ssl_keyfile or env.keyfile_path
    create_self_signed_cert = create_self_signed_cert or env.create_self_signed_cert
```

In the name of consistency, this needs to be fixed!
Also CLI > ENV

### Litestar Version

2.7.0

### Platform

- [ ] Linux
- [ ] Mac
- [ ] Windows
- [ ] Other (Please specify in the description above)

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/3188">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/3188/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by kedod**
## Description

- Added precedence of CLI parameters over envs

- [x] Add docs
- [x] Add tests for envs
- [x] Refactor

### CODE REVIEW

File: docs/usage/cli.rst [hunk 1]
```diff
@@ -95,6 +95,9 @@ The ``run`` command executes a Litestar application using `uvicorn <https://www.
 
     This feature is intended for development purposes only and should not be used to deploy production applications.
 
+.. attention::
+   CLI options take precedence over environmen variables!
```
Comment (peterschutt) [inline docs/usage/cli.rst:99]:
```suggestion
   CLI options take precedence over environment variables!
```

File: litestar/cli/commands/core.py [hunk 2]
```diff
@@ -207,20 +239,8 @@ def run_command(
...
-    port = env.port if env.port is not None else port
-    fd = env.fd if env.fd is not None else fd
-    uds = env.uds or uds
-    reload = env.reload or reload or bool(reload_dirs) or bool(reload_include) or bool(reload_exclude)
-    workers = env.web_concurrency or wc
-
-    ssl_certfile = ssl_certfile or env.certfile_path
-    ssl_keyfile = ssl_keyfile or env.keyfile_path
-    create_self_signed_cert = create_self_signed_cert or env.create_self_signed_cert
+    reload = reload or bool(reload_dir) or bool(reload_include) or bool(reload_exclude)
```
Comment (peterschutt) [inline litestar/cli/commands/core.py:242]: Can we remove all of these fields from `LitestarEnv` now?

File: tests/unit/test_cli/test_core_commands.py [hunk 3]
```diff
@@ -255,6 +254,115 @@ def test_run_command_with_app_factory(
...
+    "cli, env, expected",
+    (
+        (
+            ("--reload", True),
+            ("LITESTAR_RELOAD", True),
+            "--reload",
+        ),
+        (
+            ("--reload-dir", [".", "../somewhere_else"]),
+            ("LITESTAR_RELOAD_DIRS", [".", "../somewhere_else"]),
```
Comment (peterschutt) [inline tests/unit/test_cli/test_core_commands.py:267]: should these values be different in order to check the precedence is correct?

File: tests/unit/test_cli/test_core_commands.py [hunk 4]
```diff
@@ -255,6 +254,115 @@ def test_run_command_with_app_factory(
     )
 
 
+@pytest.mark.parametrize(
+    "cli, env, expected",
+    (
+        (
+            ("--reload", True),
+            ("LITESTAR_RELOAD", True),
```
Comment (peterschutt) [inline tests/unit/test_cli/test_core_commands.py:262]: same here, and some others.

**Review by peterschutt — CHANGES_REQUESTED** [review]
Thanks @kedod - I agree that cli args should take precedence over env vars. A few comments, LMK if I've misunderstood anything.

Cheers.

File: litestar/cli/commands/core.py [hunk 5]
```diff
@@ -120,15 +124,32 @@ def info_command(app: Litestar) -> None:
 
 
 @command(name="run")
-@option("-r", "--reload", help="Reload server on changes", default=False, is_flag=True)
-@option("-R", "--reload-dir", help="Directories to watch for file changes", multiple=True)
+@option("-r", "--reload", help="Reload server on changes", default=False, is_flag=True, envvar="LITESTAR_RELOAD")
 @option(
-    "-I", "--reload-include", help="Glob patterns for files to include when watching for file changes", multiple=True
+    "-R",
+    "--reload-dir",
```
Comment (JacobCoffee) [inline litestar/cli/commands/core.py:130]: I guess if we can do it like this we should add an option `reload-dirs` and deprecate this in 3.0?
idk.

File: litestar/cli/_utils.py
Comment (peterschutt) [inline litestar/cli/_utils.py:1]: nice!

File: litestar/cli/commands/core.py (hunk 5 above)
Comment (peterschutt) [reply litestar/cli/commands/core.py:130]: Am I right in thinking this means now we can use either `--reload-dir=./dir/a,./dir/b` and `--reload-dir=./dir/a --reload-dir=./dir/b` at the cli?

**Review by peterschutt — APPROVED** [review]
LGTM, cheers!

File: litestar/cli/commands/core.py (hunk 5 above)
Comment (JacobCoffee) [reply litestar/cli/commands/core.py:130]: 😕 not enough brain wrinkles for me to process

File: docs/usage/cli.rst [hunk 6]
```diff
@@ -95,6 +95,9 @@ The ``run`` command executes a Litestar application using `uvicorn <https://www.
 
     This feature is intended for development purposes only and should not be used to deploy production applications.
 
+.. attention::
+   CLI options take precedence over environment variables!
+
```
Comment (JacobCoffee) [inline docs/usage/cli.rst:100]:
```suggestion
.. attention:: CLI options take precedence over environment variables!

```

**Review by JacobCoffee — APPROVED** [review]

**Comment by JacobCoffee** [comment]
youre on a roll 🎉

**Review by provinzkraut — CHANGES_REQUESTED** [review]
@kedod Since you added the `CommaSeparatedPath` as a type, that would make this a feature and you'd have to target `develop` with it.

You could either do that or split that particular change into a different PR.

File: litestar/cli/commands/core.py [hunk 7]
```diff
@@ -103,6 +103,10 @@ def _run_uvicorn_in_subprocess(
     )
 
 
+class CommaSplittedPath(click.Path):
+    envvar_list_splitter = ","
```
Comment (JacobCoffee) [inline litestar/cli/commands/core.py:107]:
```suggestion
class CommaSplittedPath(click.Path):
    """A Click Path that splits the input string by commas.

    .. versionadded:: 2.8.0
    """
    envvar_list_splitter = ","
```

File: docs/usage/cli.rst [hunk 8]
```diff
@@ -95,6 +95,8 @@ The ``run`` command executes a Litestar application using `uvicorn <https://www.
 
     This feature is intended for development purposes only and should not be used to deploy production applications.
 
+.. attention:: CLI options take precedence over environment variables!
```
Comment (JacobCoffee) [inline docs/usage/cli.rst:98]:
```suggestion
.. versionchanged:: v2.8.0 
    CLI options take precedence over environment variables!
```

**Review by JacobCoffee — APPROVED** [review]

File: docs/usage/cli.rst [hunk 9]
```diff
@@ -95,6 +95,9 @@ The ``run`` command executes a Litestar application using `uvicorn <https://www.
 
     This feature is intended for development purposes only and should not be used to deploy production applications.
 
+.. versionchanged:: v2.8.0 
+    CLI options take precedence over environment variables!
```
Comment (JacobCoffee) [inline docs/usage/cli.rst:99]:
```suggestion
.. versionchanged:: 2.8.0
    CLI options take precedence over environment variables!
```

File: litestar/cli/commands/core.py [hunk 10]
```diff
@@ -103,6 +103,14 @@ def _run_uvicorn_in_subprocess(
     )
 
 
+class CommaSplittedPath(click.Path):
+    """A Click Path that splits the input string by commas.
+
+    .. versionadded:: 2.8.0
+    """
+    envvar_list_splitter = ","
```
Comment (JacobCoffee) [inline litestar/cli/commands/core.py:111]:
```suggestion
class CommaSplittedPath(click.Path):
    """A Click Path that splits the input string by commas.

    .. versionadded:: 2.8.0
    """

    envvar_list_splitter = ","
```

**Pull request merged**